corgidrp/calibrate_nonlin.py

```diff
@@ -0,0 +1,806 @@
+        frames in stack_arr in the same order. The sum of elements of len_list 
+        must equal to the number of sub-stacks in stack_arr. The number of 
+        elements (= the number of unique em gain values) in len_list must be 
+        one or greater.
+    
+    stack_arr2 : array-like
+        stack_arr2 is a stack array of EXCAM unity EM gain illuminated pupil L1 
+        SCI frames. stack_arr2 contains a stack of frames of uniform exp time, 
+        such that the mean signal in the pupil regions is a few thousand DN.
+    
+    actual_gain_arr : array-like
+        The array of actual EM gain values (as opposed to commanded EM gain) 
+        corresponding to the number of EM gain values used to construct 
+        stack_arr and in the same order. Note: calibrate_nonlin does not 
+        calculate actual EM gain values -- they must be provided in this array. 
+        The length of actual_gain_arr must equal the length of len_list. Values 
+        must be >= 1.0.
+        
+    norm_val : int
+        Value in DN to normalize the nonlinearity values to. Must be greater than 
+        0 and must be divisible by 20 without remainder. (1500 to 3000 recommended).
+    
+    min_write : float
+        Minimum mean value in DN to output in nonlin_arr and csv_lines. 
+        (800.0 recommended)
+    
+    max_write : float
+        Maximum mean value in DN to output in nonlin_arr and csv_lines. 
+        (10000.0 recommended)
+    
+    mkplot : boolean
+        Option to display plots. Default is None. If mkplot is anything other 
+        than None, then this option is chosen.
+        
+    verbose : boolean
+        Option to display various diagnostic print messages. Default is None. 
+        If verbose is anything other than None, then this option is chosen.
+    
+    Returns
+    -------
+    headers: array-like
+        1-D array of headers used to build csv-lines. The length is equal to 
+        the number of columns in 'nonlin_arr' and is one greater than the 
+        length of 'actual_gain_arr'.
+    
+    nonlin_arr: array-like
+        2-D array with nonlinearity values for input signal level (DN) in rows 
+        and EM gain values in columns. The input signal in DN is the first column. 
+        Signal values start with min_write and run through max_write in steps 
+        of 20 DN.
+    
+    csv_lines : list
+        List of strings containing the contents of 'headers' and 'nonlin_arr'.
+    
+    means_min_max : array-like
+        minima and maxima of mean values (in DN) used the fit each for EM gain. 
+        The size of means_min_max is N x 2, where N is the length of actual_gain_arr.
+    """
+    # copy stack_arr and stack_arr2 and cast them into np arrays for convenience
+    stack_arr, stack_arr2 = copy_and_cast(stack_arr, stack_arr2)    
+
+    # input checks
+    # load in default parameters
+    from corgidrp.detector import nonlin_params as master_files
```

Comment:
I think this should be at the top.

Reply:
@maxwellmb @semaphoreP: what is the syntax for docstrings (Args) for calibration functions in corgidrp accepting **kwargs? (I only found walker using it and there's no docstring associated with that particular use)

Reply:
you'd want the function to look _something_ like: 
`def calibrate_kgain(input_dataset, logplot1=-1, logplot2=4, logplot3=200)`
(as a very basic example) where I'm grabbing `logplot1` from the kgain_params.

Reply:
@maxwellmb: Okay, when saying "be kwargs", you mean explictly writing any keyword in the definition, not using **kwargs: `def calibrate_kgain(input_dataset, **kwargs)`?

Reply:
yeah! :D

Reply:
ty

Reply:
@JuergenSchreiber: FYI, I sent an email to Guillermo Gonzalez and Nathan Bush, CC'ed Max, asking which of each of the kgain and nonlin params are constant or variable.

Reply:
Guillermo's reply for both kgain and nonlin parameters: "the only parameters that might need adjusted are the 'rms_low_limit' and 'rms_upp_limit' for nonlin, which were determined with simulated frames."

Reply:
@JuergenSchreiber @maxwellmb After reviewing the list with Guillermo in more detail, these are the parameters that may be keywords in the Kgain and nonlin functions:  **'logplot1', 'logplot2', ‘logplot3’, lowess_frac’, 'rms_low_limit' and 'rms_upp_limit'.**

I'll ask again Guillermo on Thu 07/11 to confirm if there are any changes to this list.

Reply:
Guillermo confirms as of 07/11/24 that these are the parameters in Kgain and Nonlin that may vary along the mission: **'logplot1', 'logplot2', ‘logplot3’, ‘lowess_frac’, 'rms_low_limit', 'rms_upp_limit’, 'fit_upp_cutoff1', 'fit_upp_cutoff2', 'fit_low_cutoff1' and 'fit_low_cutoff2'.**

Reply:
Ok great. These look like appropriate keyword arguments (but maybe with clearer names).